### PR TITLE
Implement analytics in Messenger and Toggler

### PR DESCRIPTION
**Pull request by joeyyang**
This PR makes the following changes:
- Adds analytics tracking via Microsoft Application Insights to `Messenger` class. 
- Implements analytics in Toggler, logging toggle commands.

Example event format:
Screen Shot 2021-12-09 at 3 57  00PM

Example aggregation view:
Screen Shot 2021-12-09 at 3 57  12PM

Questions:
- How should analytics be implemented in each plugin? What is the API for calling analytics?
- Where should keys go?
- Is `Messenger` the best place for tracking to live? Should `stats` be its own package?

### CODE REVIEW

**Comment by lynnerang** [comment]
This is excitingggg! :D 

Might be missing some context from standup and see this is a WIP, but at a glance:
- noticed a few comments in the code, but looks like they might be intentional placeholders while you work out some deets
- RE where it should be implemented and if it belongs in Messenger, I'm curious where we tend to want to log stats throughout the plugins e.g.:
    - if there's usually a toast and/or console log in each instance we want to log a stat seems like it goes especially nicely with Messenger
    - if the action name and payload in app thread messages tends to be the same info we want to include in the stats, maybe it'd be an option to have a catch-all stats log in the code.ts message handling switch that uses those

just a few thoughts/hypotheticals that came to mind, but if you think it'd be fun, maybe we could fold this into the experimentation with logging in Specter 2.0 since there's a lot we could record and it might help hone in on some possible patterns?

**Comment by joeyyang** [comment]
hey @lynnerang thanks for your review! re: your feedback:
- i think you're right that we could log every `action` that passes through the standard `dispatcher` / `runAction` command in the plugins' `code.ts` - are all the `action`s in there generally events we want to track? if so, i believe we can subdivide those larger `action`s by `payload` data
- yes, i'd love to roll this into Specter 2.0 - how would you like to do this?

File: toggler/src/App.ts [hunk 1]
```diff
@@ -573,6 +577,16 @@ export default class App {
 
     instancesToToggle.forEach(toggleInstance);
 
+    // messenger.track(`Selection toggled to ${toggleCommand.targetSchemeName}, ${toggleCommand.targetThemeName}`);
+    figma.ui.postMessage({
```
Comment (gblakeman) [inline toggler/src/App.ts:581]: Should this block get added to `Messenger` in some way so it’s even easier to invoke?

**Review by gblakeman — COMMENTED** [review]
I like where this is headed! Left a few questions inline.

I continue to be in favor of leaving as much of this as possible in `Messenger` since it’s already something we drop into each plugin anyway, and to @lynnerang’s point we will likely want to track something we are already logging.

Personally, I don’t think it makes sense to have this simply log every `action` run through dispatcher. I think that could lead to a lot of data that isn’t meaningful (or is hard to parse). I think we’ll want the ability to start tracking things on a smaller scale (explicit events) and build it up over time in each of the plugins.

One thing I would like to see still is an example of a tracking event that was initiated on the UI-side, just so we can think through the workflow for this. Or, we might decide that we only want to fire off a tracking event from the App thread, based on what was received from the UI thread?

File: toggler/.env.example [hunk 2]
```diff
@@ -5,3 +5,18 @@ SECRET_KEY=xxxx1234xxxx1234xxxx1234
 
 // URL of Git repo to import Mercado token reference from
 MERCADO_TOKEN_REPO_URL="http://www.example.com/repo/url"
+
+// Unique keys for other plugins in the library
+// These must match the keys used in each plugin’s respective .env file
+SECRET_KEY_EXONYM=xxxx1234xxxx1234xxxx1234xxxx1234
+SECRET_KEY_REALISH=xxxx1234xxxx1234xxxx1234xxxx1234
+SECRET_KEY_SPECTER=xxxx1234xxxx1234xxxx1234xxxx1234
```
[the remark this answers is not in the document]
Comment (joeyyang) [reply toggler/.env.example:13]: nope! i'll remove

File: toggler/src/App.ts (hunk 1 above)
Comment (joeyyang) [reply toggler/src/App.ts:581]: Does `Messenger` have access to `figma.ui`? (Does it depend on where `Messenger` is instantiated?) If so, I think that'd be great

**Comment by joeyyang** [comment]
thanks @gblakeman! re: your comments:

> Personally, I don’t think it makes sense to have this simply log every action run through dispatcher. I think that could lead to a lot of data that isn’t meaningful (or is hard to parse). I think we’ll want the ability to start tracking things on a smaller scale (explicit events) and build it up over time in each of the plugins.
- I agree that we probably won't find tracking data about every app action useful. That said, with custom dashboard views, irrelevant events should be easy enough to hide. As we discussed though, it could be a nice pattern to track events in the plugins' `dispatcher` > `runAction` function, maybe like this - does this make sense as a way to call the tracking function?
```ts
  const runAction = async () => {
    switch (type) {
      case 'toggle-theme':
        App.toggleTheme(payload);
        Messenger.track(case, 'Toggle Theme', payload);
      break;
      ...
    }
  }
```
> One thing I would like to see still is an example of a tracking event that was initiated on the UI-side, just so we can think through the workflow for this. Or, we might decide that we only want to fire off a tracking event from the App thread, based on what was received from the UI thread?
- I think it'd look something like this:
```html
<script>
    parent.postMessage({
      pluginMessage: {
        action: 'track',
        payload: {
          eventName: 'event name',
          customProperties: {
            ...,
          },
        },
      },
    }, '*');
</script>
```
We don't currently instantiate any `Messenger` instances from the UI, do we?

File: plugin-tools/src/Messenger.ts [hunk 3]
```diff
@@ -48,6 +68,65 @@ export default class Messenger {
...
+      action: 'track',
+      payload: {
+        eventName,
+        eventData,
+        eventType,
+      },
+    });
+  }
+
+
```
Comment (gblakeman) [inline plugin-tools/src/Messenger.ts:95]: Extra line

File: toggler/.env.example [hunk 4]
```diff
@@ -5,3 +5,6 @@ SECRET_KEY=xxxx1234xxxx1234xxxx1234
 
 // URL of Git repo to import Mercado token reference from
 MERCADO_TOKEN_REPO_URL="http://www.example.com/repo/url"
+
+// Microsoft Application Insights instrumentation key
+APP_INSIGHTS_CONNECTIONSTRING="InstrumentationKey=l0l0l0l0-l0l0-l0l0-l0l0-l0l0l0l0l0l0;IngestionEndpoint=https://your-region.in.applicationinsights.azure.com/"
```
Comment (gblakeman) [inline toggler/.env.example:10]: To me it would make more sense to split this and build the proper string in `Messenger`. I think it would be less error-prone.

```
APP_INSIGHTS_KEY=l0l0l0l0-l0l0-l0l0-l0l0-l0l0l0l0l0l0
APP_INSIGHTS_ENDPOINT=your-region.in.applicationinsights.azure.com
```

```js
const appInsightsKey = `InstrumentationKey=${process.env.APP_INSIGHTS_KEY};IngestionEndpoint=https://${process.env.APP_INSIGHTS_ENDPOINT}/`
```

File: toggler/src/GUI.ts [hunk 5]
```diff
@@ -1,5 +1,7 @@
+import { Messenger } from 'linkedin-design-tools-figma';
```
Comment (gblakeman) [inline toggler/src/GUI.ts:1]: Can you fully remove the local `Messenger` class file from Toggler now? Or is it being used elsewhere?

**Review by gblakeman — APPROVED** [review]
🌈🦄

Nice work! Excited to start playing with this!

My only minor quibble is with the order of the functions in `Messenger` – should we keep `log` and `error` next to each other since they’re both for local consoles and then put `track` above or below them with `sendEvent`?

File: toggler/.env.example (hunk 4 above)
Comment (joeyyang) [reply toggler/.env.example:10]: the key comes in that format from the dashboard copy/paste result - figured folks are just gonna paste over the whole thing but wanted to still show syntax & what string parts to expect

File: plugin-tools/src/Messenger.ts [hunk 6]
```diff
@@ -48,6 +68,65 @@ export default class Messenger {
...
+  /**
+   * Like `log` above, but also logs the event to Microsoft Application Insights.
+   * Can only run when Messenger is initialized through the iframe environment, not the main app.
+   * If `appInsightsKey` is not given at Messenger instance construction, still logs to console.
+   * Event reporting delay SLA is below:
+   *   https://azure.microsoft.com/en-us/support/legal/sla/application-insights/v1_0/
```
[the remark this answers is not in the document]
Comment (joeyyang) [reply plugin-tools/src/Messenger.ts:101]: nice, thank you!

File: toggler/src/GUI.ts (hunk 5 above)
Comment (joeyyang) [reply toggler/src/GUI.ts:1]: below on line 24, creates a `messenger` instance to log the event to server

Comment (joeyyang) [reply toggler/src/GUI.ts:1]: did you mean we should call `messenger` somewhere else?

File: toggler/.env.example (hunk 4 above)
Comment (gblakeman) [reply toggler/.env.example:10]: Ah, yeah, I wondered. In that case, this is perfect!

File: toggler/src/GUI.ts (hunk 5 above)
Comment (gblakeman) [reply toggler/src/GUI.ts:1]: I mean the actual Messenger.ts file that is in `/toggler/src`. If you’re now using the one from the shared tools package, the local one shouldn’t be necessary, right?

Comment (joeyyang) [reply toggler/src/GUI.ts:1]: OH yes we can thanks!
